File: python/phonebook.py
from typing import List, Tuple


class Phonebook:
    def __init__(self):
        self.storage = {}

    def add(self, name, number):
        self.storage[name] = number

    def lookup(self, name):
        return self.storage[name]

    def __len__(self):
        return len(self.storage)

    def find_clashes(self, number_to_check) -> List[Tuple[str, str]]:
        entries = []
        for name, number in self.storage.items():
            if number.startswith(number_to_check) or number_to_check.startswith(number):
                entries.append((name, number))
        return entries

File: csharp/EnterprisePhonebook-TestDesign-Kata/EnterprisePhonebook/Phonebook.cs
namespace SammanCoaching.Phonebook;

public sealed class Phonebook
{
    private readonly Dictionary<string, string> _storage = new();

    public void Add(string name, string number) => _storage[name] = number;

    public string Lookup(string name) => _storage[name];

    public int Count => _storage.Count;

    public IReadOnlyList<KeyValuePair<string, string>> FindClashes(string numberToCheck) =>
        _storage
            .Where(entry => entry.Value.StartsWith(numberToCheck) || numberToCheck.StartsWith(entry.Value))
            .ToList();
}
